File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/ServicesViewController.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSSharingServicePickerTouchBarItem in an NSTouchBar instance.
*/

#import "ServicesViewController.h"

static NSTouchBarCustomizationIdentifier SharingCustomizationIdentifier = @"com.TouchBarCatalog.sharingViewController";
static NSTouchBarItemIdentifier ServicesItemIdentifier = @"com.TouchBarCatalog.sharing";

@interface ServicesViewController () <NSTouchBarDelegate, NSSharingServicePickerTouchBarItemDelegate, NSSharingServiceDelegate>

@property (strong) NSSharingServicePickerTouchBarItem *servicesTouchBarItem;
@property (strong) NSImage *imageToSend;

@end


#pragma mark -

@implementation ServicesViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    _imageToSend = [NSImage imageNamed:@"image3"];
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = SharingCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers =
        @[ServicesItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[ServicesItemIdentifier];
    
    bar.principalItemIdentifier = ServicesItemIdentifier;
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:ServicesItemIdentifier])
    {
        if (self.servicesTouchBarItem == nil)
        {
            _servicesTouchBarItem = [[NSSharingServicePickerTouchBarItem alloc] initWithIdentifier:identifier];
            self.servicesTouchBarItem.delegate = self;
        }
        return self.servicesTouchBarItem;
    }
    
    return nil;
}


#pragma mark - NSSharingServicePickerTouchBarItemDelegate

- (NSArray *)itemsForSharingServicePickerTouchBarItem:(NSSharingServicePickerTouchBarItem *)pickerTouchBarItem
{
    return @[self.imageToSend];
}


#pragma mark - NSSharingServiceDelegate

- (id<NSSharingServiceDelegate>)sharingServicePicker:(NSSharingServicePicker *)sharingServicePicker delegateForSharingService:(NSSharingService *)sharingService
{
    return self;
}

- (NSRect)sharingService:(NSSharingService *)sharingService sourceFrameOnScreenForShareItem:(id)item
{
    return NSMakeRect(0, 0, self.imageToSend.size.width, self.imageToSend.size.height);
}

- (NSWindow *)sharingService:(NSSharingService *)sharingService sourceWindowForShareItems:(NSArray *)items sharingContentScope:(NSSharingContentScope *)sharingContentScope
{
    return [self.view window];
}

- (NSImage *)sharingService:(NSSharingService *)sharingService transitionImageForShareItem:(id)item contentRect:(NSRect *)contentRect
{
    return (NSImage *)item;
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/ServicesViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSSharingServicePickerTouchBarItem in an NSTouchBar instance.
*/

import Cocoa

class ServicesViewController: NSViewController {
    let servicesBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.servicesBar")
    let services = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.services")
    
    var imageToSend = NSImage(named: "image3")!
    
    // MARK: - NSTouchBarProvider
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        
        touchBar.customizationIdentifier = servicesBar
        touchBar.defaultItemIdentifiers = [services]
        touchBar.customizationAllowedItemIdentifiers = [services]
        touchBar.principalItemIdentifier = services
        
        return touchBar
    }
}

// MARK: - NSTouchBarDelegate

extension ServicesViewController: NSTouchBarDelegate {
    
    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        guard identifier == services else { return nil }
        
        let services = NSSharingServicePickerTouchBarItem(identifier: identifier)
        services.delegate = self
        
        return services
    }
}

// MARK: - NSSharingServicePickerTouchBarItemDelegate

extension ServicesViewController: NSSharingServicePickerTouchBarItemDelegate {
    func items(for pickerTouchBarItem: NSSharingServicePickerTouchBarItem) -> [Any] {
        return [imageToSend]
    }
    
}

// MARK: - NSSharingServiceDelegate

extension ServicesViewController: NSSharingServiceDelegate {
    
    // MARK: - NSSharingServicePickerDelegate
    
    func sharingServicePicker(_ sharingServicePicker: NSSharingServicePicker,
                              delegateFor sharingService: NSSharingService) -> NSSharingServiceDelegate? {
        return self
    }
    
    // MARK: - NSSharingServiceDelegate
    
    func sharingService(_ sharingService: NSSharingService, sourceFrameOnScreenForShareItem item: Any) -> NSRect {
        return NSRect(x: 0, y: 0, width: imageToSend.size.width, height: imageToSend.size.height)
    }
    
    func sharingService(_ sharingService: NSSharingService,
                        transitionImageForShareItem item: Any, contentRect: UnsafeMutablePointer<NSRect>) -> NSImage? {
        return imageToSend
    }
    
    public func sharingService(_ sharingService: NSSharingService,
                               sourceWindowForShareItems items: [Any],
                               sharingContentScope: UnsafeMutablePointer<NSSharingService.SharingContentScope>) -> NSWindow? {
        return view.window
    }
    
}

